### PR TITLE
Optimizations for the JS version

### PR DESCRIPTION
**Pull request by MatmaRex**
See commit messages, and my comment here for context: https://qntm.org/wasm#komment673abf35388f2

### CODE REVIEW

File: examples/perf.js [hunk 1]
```diff
@@ -19,16 +21,28 @@ const d = xys.map(([xu, yu]) =>
...
+const cycleWasm = await getCycleWasm(d);
 console.timeEnd('HK/WASM')
+if (!quiet) {
+  console.log(cycleWasm)
+}
 console.log()
+
+if (JSON.stringify(cycleJs) !== JSON.stringify(cycleWasm)) {
+  console.log('RESULT MISMATCH!')
+}
```
Comment (qntm) [inline examples/perf.js:48]: Please could you leave the performance testing script unmodified for now. But it looks like there would be some value in adding a unit test to the Mocha test suite which generates a 2D TSP problem and ensures that the JavaScript and WebAssembly implementations generate precisely identical results?

File: src/js-impl/index.js [hunk 2]
```diff
@@ -22,17 +23,15 @@ export const getCycle = d => {
 
     If `S` contains only a single city, `u`, then `len[S][u]` is `d[n - 1][u]`.
   */
-  const len = []
+  const len = new Float64Array(all * n)
```
Comment (qntm) [inline src/js-impl/index.js:26]: `S` can vary from 1 to 2<sup>*n* - 1</sup> - 1 inclusive. We leave space in the array for `S` = 0 as well although we don't use it. So in practice `S` ranges from 0 to 2<sup>*n* - 1</sup> - 1 inclusive, a range of 2<sup>*n* - 1</sup> possible values. Meanwhile, `u` ranges from 0 to *n* - 2 inclusive, a range of *n* - 1 possible values. So this should be `new Float64Array((1 << (n - 1)) * (n - 1))`. Compare with [what we do for the WebAssembly](https://github.com/qntm/held-karp/blob/1d2d1d6720d5db9baf3523c9a33f1068e62377bf/src/wasm-impl/index.js#L18-L19).

(`all * n` does work, as it is always larger than or equal to `(1 << (n - 1)) * (n - 1)`, but it allocates a little more memory than necessary.)

Likewise for `prev`, below.

This is, however, a great optimisation, can't believe I didn't think of it! 👍

File: src/js-impl/index.js [hunk 3]
```diff
@@ -41,8 +40,8 @@ export const getCycle = d => {
       const S2 = S ^ (1 << v)
       // Is v in S?
       if (S2 < S) {
-        let bestL
-        let bestU
+        let bestL = 0
+        let bestU = 0
```
Comment (qntm) [inline src/js-impl/index.js:44]: I learned something! 👍

**Review by qntm — CHANGES_REQUESTED** [review]
Extremely good performance enhancements, only a few small requested modifications.

**Comment by qntm** [comment]
It also looks like this change circumvents the crash I was seeing when using very large `Array`s and allows us to go to more than 23 cities with the JavaScript implementation. I've had it working up to 28 cities - it chewed up all of my RAM and took nearly 7 minutes but it worked.

File: examples/perf.js (hunk 1 above)
Comment (MatmaRex) [reply examples/perf.js]: No problem, undid these.

**Review by qntm — APPROVED** [review]

**Pull request merged**